sycl/include/CL/sycl/detail/kernel_program_cache.hpp

```diff
@@ -0,0 +1,95 @@
+//==--- kernel_program_cache.hpp - Cache for kernel and program -*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/locked.hpp>
+#include <CL/sycl/detail/os_util.hpp>
+#include <CL/sycl/detail/pi.hpp>
+
+#include <atomic>
+#include <condition_variable>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+
+namespace cl {
+namespace sycl {
+namespace detail {
+class kernel_program_cache {
+public:
+  /// Denotes pointer to some entity with its build state.
+  /// The pointer is not null if and only if the entity is successfully built.
+  /// Build state is provided by the user of cache instance.
+  /// Currently there is only a single user - ProgramManager class.
+  template<typename T>
+  struct EntityWithBuildState {
+    std::atomic<T *> Ptr;
+    std::atomic<int> State;
+
+    EntityWithBuildState(T* P, int S)
+      : Ptr{P}, State{S}
+    {}
+  };
+
+  using PiProgramT = std::remove_pointer<RT::PiProgram>::type;
+  using PiProgramPtrT = std::atomic<PiProgramT *>;
+  using ProgramWithBuildStateT = EntityWithBuildState<PiProgramT>;
+  using ProgramCacheT = std::map<OSModuleHandle, ProgramWithBuildStateT>;
+
+  using PiKernelT = std::remove_pointer<RT::PiKernel>::type;
+  using PiKernelPtrT = std::atomic<PiKernelT *>;
+  using KernelWithBuildStateT = EntityWithBuildState<PiKernelT>;
+  using KernelByName = std::map<string_class, KernelWithBuildStateT>;
+  using KernelCacheT = std::map<RT::PiProgram, KernelByName>;
+
+  ~kernel_program_cache();
+
+  locked<ProgramCacheT> acquireCachedPrograms() {
+    return {m_CachedPrograms, m_ProgramCacheMutex};
+  }
+
+  // Non thread-safe
+  ProgramCacheT &getCachedPrograms() {
+    return m_CachedPrograms;
+  }
+
+  locked<KernelCacheT> acquireKernelsPerProgramCache() {
+    return {m_KernelsPerProgramCache, m_KernelsPerProgramCacheMutex};
+  }
+
+  // Non thread-safe
+  KernelCacheT &getKernelsPerProgramCache() {
+    return m_KernelsPerProgramCache;
+  }
+
+  template<class Predicate>
+  void waitUntilBuilt(Predicate Pred) {
+    std::unique_lock<std::mutex> Lock(m_BuildCVMutex);
+
+    m_BuildCV.wait(Lock, Pred);
+  }
+
+  void notifyAllBuild() {
+    m_BuildCV.notify_all();
+  }
+
+private:
+  std::mutex m_ProgramCacheMutex;
+  std::mutex m_KernelsPerProgramCacheMutex;
+
+  std::condition_variable m_BuildCV;
+  std::mutex m_BuildCVMutex;
+
+  ProgramCacheT m_CachedPrograms;
```

Comment:
We are trying to follow LLVM coding guidelines(again).
It says variable name should start for capital letter. (In some places we use MVarName for members)